frontend/src/modules/dataset/dataset-browser.tsx

```diff
@@ -0,0 +1,389 @@
+    const isSelectingPrefix = selectableItemsTypes.indexOf('prefixes') > -1;
+    const [state, setState] = React.useReducer(datasetBrowserReducer, {
+        datasetContext: undefined,
+        breadcrumbs: [],
+        items: [],
+        selectedItems: [],
+        manageMode,
+        isLoading: false,
+        nextToken: undefined,
+        username: username!,
+        projectName: projectName!,
+        filter: {
+            filteringText: '',
+            filteringTextDisplay: '',
+            filteredItems: []
+        },
+        pagination: {
+            currentPageIndex: 1,
+            pagesCount: 1,
+            openEnd: false,
+            disabled: false,
+        },
+        refresh: false
+    });
+
+    const displayMode = displayModeForDatasetContext(state.datasetContext);
+    const isCreating = state.manageMode === DatasetBrowserManageMode.Create;
+
+    // memoize notification service so it doesn't cause infinite loops for fetchDatasetResources & fetchDatasets
+    const notificationService = useMemo(() => {
+        return NotificationService(dispatch);
+    }, [dispatch]);
+
+    const fetchDatasetResources = useMemo(() => (datasetContext: Partial<DatasetContext>, nextToken?: string, existingItems?: (DatasetResource | IDataset)[]) => {
+        if (state.manageMode !== DatasetBrowserManageMode.Create) {
+            return dispatch(getDatasetContents({datasetContext, projectName, username, nextToken})).then((response) => {
+                if (isFulfilled(response)) {
+                    // if nextToken is provided we're appending to existing items
+                    const items = nextToken ? [...existingItems || []] : [];
+        
+                    const payload: Partial<Pick<DatasetBrowserState, 'items' | 'nextToken' | 'datasetContext' | 'filter' | 'isLoading' | 'refresh'>> = {
+                        items: items.concat(response.payload.data.contents),
+                        nextToken: response.payload.data.nextToken,
+                        isLoading: false
+                    };
+        
+                    // if the dataset context location is for a resource update the filter too
+                    const resourceStringForPath = resourceForPath(datasetContext.location);
+                    if (resourceStringForPath) {
+                        payload.filter = {
+                            filteredItems: [],
+                            filteringTextDisplay: resourceStringForPath,
+                            filteringText: resourceStringForPath
+                        };
+                    }
+        
+                    setState({
+                        type: DatasetActionType.State,
+                        payload
+                    });
+                } else {
+                    // if an error occurred reset the component to a safe state
+                    notificationService.generateNotification(`Failed to fetch resources for ${datasetContext.name}`, 'error');
```

Comment:
Do we currently provide users a detailed reason why a dataset fetch failed? If not then there may be an existing reason, so this would be worth bringing up with the team. 

Also do we currently return specific error messages that could be used in the frontend? If not then this seems out of scope as this code is mimicking existing behavior.

Reply:
There isn't really anything other to do than retry. Either the backend failed randomly (not sure what would cause this) or they queried for an invalid dataset. Either way the datasetContext is reset and they're taken back to the default DatasetBrowser state so they can try again.